### DISCUSSION TITLE
qpdf does not recognize some font types

### DISCUSSION THREAD

**Discussion by aminreguig** [qpdf/qpdf#984] [Q&A] [answered]
I have an pdf file that I want to make some modifications by converting it to qdf, do modifications and re convert it to pdf :
```
qpdf -qdf Convocation.pdf output.qdf

#modifucations

fix-qdf < output.qdf > final.pdf
```

But I find that some text writen with a soecific font type (Averta) does not appear. I think that this code does not support all font types. 
thank you for suggesting some solutions

**Comment by jberkenbilt** [comment]
It is very common (the norm, in fact) for PDF files to contain subsetted fonts. The only glyphs from a font that are included in the PDF are typically those corresponding to the characters that are actually present. So if you are editing text and expecting your edits to work, you would have to stick with characters that already appear in the file. If you use a character whose glyph is not embedded, it will not display. qdf doesn't work very well for editing text except in the simplest of cases because PDF files' representation of text is to place specific groups of characters in exact places on the page. If you have a simple, hand-generated PDF that uses built-in fonts or fonts that were embedded in their entirety and the thing that generated the PDF is not doing anything fancy with kerning or justification, you might be lucky, but generally this approach won't work. If you are trying to edit text, you will need to use a tool that knows how to render the PDF and can have access to the original fonts.

If this explanation explains what you're seeing, great. If not, please provide more details, and maybe I can help. But either way, QDF mode provides a low-level way to work with PDF files. You have to have some familiarity with the PDF spec to use it effectively.

**Reply by aminreguig** [reply]
thank you @jberkenbilt  for your message.

No, I did not change any text. I just changed some colors from RGB to CMYK. 

When I reconvert the qdf file to a pdf file, I loose all texts written with a font type downloaded from internet called Averta (It does not exist by default in the system) but I dont loose the rest  texts writen with Times and Arial.

**Reply by jberkenbilt** [reply]
Are you using fix-qdf afterwards?  I am not aware of any way that going through a qdf cycle could do this. If you just run qpdf on the file, does the output still work? If not, there could be something going on involving the file talking to an outside service and encryption or something. If running qpdf on the file (qpdf in.pdf out.pdf, then check out.pdf) works but editing the qdf does not work and you are running fix-qdf so that `qpdf --check` is not returning any errors or warnings, then my guess is that you are making some error in editing the PDF. If it is a file you are able to share, I can take a look. If you are not able to share it publicly but are able to share it privately, you can send me the file by email with a description of what you're trying to do and I can try a sample change just enough to see if the font is damaged or determine whether there's something about the way the font is referenced that is causing an issue. That will probably either expose user error on your part and hopefully help you figure out the right approach or teach me something new about different ways of handling fonts that I am not currently aware of.

**Reply by aminreguig** [reply]
I changed only some RGB colors with CMYK colors by replacing for example "1 1 1 rg" with "0 100 100 0 k"

And yes, can you give me your email to send you the pdf file ?

**Reply by jberkenbilt** [reply]
Please be sure to provide specific information about your change, e.g.
* ran qpdf --qdf infile.pdf outfile.pdf
* on line 12345, changed "1 1 1 rg" to "0 100 100 0 k"
* ran fix-qdf < outfile.pdf > fixed.pdf
* observe on page 6 that the words "xyz" on the left hand side a third of the way down disappeared but were there on the original

etc. Don't assume I will notice the difference without it being pointed out (though I will run an image diff). Also would help if you indicate what viewer you are using and whether you have to do anything (like highlight text) to see the difference. The less time I have to spend being a detective on the procedure, the more time I can spend being a detective on the root cause. :-) Thanks!

**Comment by jberkenbilt** [comment]
After having looked at some files send by private email, it seems that the qdf file was edited in a text editor that isn't safe to use with binary files, so the non-ASCII characters in the embedded font file got corrupted as the editor tried to interpret them as Unicode.

**Reply by jberkenbilt** [reply, marked as answer]
I am not able to advise on editing binary files in Windows, but I provided an alternate approach:
* Use `qpdf --show-pages` to get the object ID of the content stream. In this case, it was 4.
* Use qpdf --json to create an update file for that object:
  ```
  qpdf --decode-level=generalized --json-output infile.pdf temp.json --json-stream-data=file --json-object=4
  ```
  This creates `temp.json` and `temp.json-4`. The `temp.json-4` file is the content stream as a stand-alone text file. This can be edited safely in a non-binary-aware text editor. The PDF file allows Windows-style newlines, so it's even okay if the editor changes the line endings.
* Insert the updated stream into a copy of the file:
   ```
   qpdf infile.pdf fixed.pdf --update-from-json=temp.json
   ```

The resulting file has the edited content stream.

**Reply by aminreguig** [reply]
It works well. Thank you very much for your help
